pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -110,37 +112,59 @@ lazy_static! {
 }
 
 lazy_static! {
-    static ref SYNC_QUEUE: Mutex<BinaryHeap<SyncTask>> = Mutex::new(BinaryHeap::new());
+    static ref SYNC_QUEUE: SegQueue<SyncTask> = SegQueue::new();
 }
 
-/// An image sync task to store in the priority queue.
-/// The task priority is defined by its [`PartialOrd`] derive:
-/// * lower enum variants are of more priority compared to the higher ones
-/// * for the same enum variants, "natural" comparison happens for their data
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
-enum SyncTask {
+/// A task to run in the async download/upload loop.
+/// Limited by the number of retries, after certain threshold the failing task gets evicted and the timeline disabled.
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct SyncTask {
+    retries: u32,
+    kind: SyncKind,
+}
+
+impl SyncTask {
+    fn new(retries: u32, kind: SyncKind) -> Self {
+        Self { retries, kind }
+    }
+
+    fn tenant_id(&self) -> ZTenantId {
+        match &self.kind {
+            SyncKind::Download(download_kind) => download_kind.tenant_id,
+            SyncKind::Upload(upload_kind) => upload_kind.tenant_id,
+        }
+    }
+
+    fn timeline_id(&self) -> ZTimelineId {
+        match &self.kind {
+            SyncKind::Download(download_kind) => download_kind.timeline_id,
+            SyncKind::Upload(upload_kind) => upload_kind.timeline_id,
+        }
+    }
```

Comment:
Should we just pull these into SyncTask

Reply:
That's quite a change actually, plenty of things are based on that.
I have a similar idea to introduce a `pub struct TimelineSyncId(ZTenantId, ZTimelineId);`, will try to bring it into with your suggestion.

Reply:
Done.